src/CredentialsWrapper.php

```diff
@@ -167,19 +178,38 @@ public function getBearerString()
     }
 
     /**
+     * @param string $audience optional audience for self-signed JWTs.
      * @return callable Callable function that returns an authorization header.
      */
-    public function getAuthorizationHeaderCallback()
+    public function getAuthorizationHeaderCallback($audience = null)
     {
         $credentialsFetcher = $this->credentialsFetcher;
         $authHttpHandler = $this->authHttpHandler;
 
         // NOTE: changes to this function should be treated carefully and tested thoroughly. It will
         // be passed into the gRPC c extension, and changes have the potential to trigger very
         // difficult-to-diagnose segmentation faults.
-        return function () use ($credentialsFetcher, $authHttpHandler) {
-            $token = self::getToken($credentialsFetcher, $authHttpHandler);
-            return empty($token) ? [] : ['authorization' => ["Bearer $token"]];
+
+        return function (array $headers = []) use ($credentialsFetcher, $authHttpHandler, $audience) {
+            $token = $credentialsFetcher->getLastReceivedToken();
+            if (self::isExpired($token)) {
+                // Call updateMetadata to take advantage of self-signed JWTs
+                if ($credentialsFetcher instanceof UpdateMetadataInterface) {
+                    return $credentialsFetcher->updateMetadata($headers, $audience);
+                }
```

Comment:
Although this looks like a lot, I did try to heed the comment above regarding segfaults. This code only changese the previous behavior to use `updatMetadata` instead of `fetchAuthToken` if the fetcher allows it. And this only has the effect of using self-signed JWTs when available.

This SHOULD be tested with protobuf/grpc extensions enabled! Here is the code I've written to do this, although I do not see a good way in the existng test suites to verify this behavior:

```php
require __DIR__ . '/vendor/autoload.php';

use Google\ApiCore\Call;
use Google\ApiCore\CredentialsWrapper;
use Google\ApiCore\Transport\GrpcTransport;
use Google\ApiCore\Testing\MockRequest;
use Grpc\ChannelCredentials;

$opts = ['credentials' => ChannelCredentials::createInsecure()];
$wrapper = CredentialsWrapper::build();
$transport = new GrpcTransport(
    $hostname = 'storage.googleapis.com',
    $opts
);

$call = new Call(
    'method',
    null,
    new MockRequest()
);

$transport->startUnaryCall($call, [
    'audience' => 'foo.com',
    'credentialsWrapper' => $wrapper
])->wait();
```